staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/promise/promise.go

```diff
@@ -0,0 +1,112 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package promise
+
+import (
+	"fmt"
+	"sync"
+
+	"k8s.io/apiserver/pkg/util/flowcontrol/counter"
+)
+
+// promise implements the promise.WriteOnce interface.
+// This implementation is based on a condition variable.
+// This implementation tracks active goroutines:
+// the given counter is decremented for a goroutine waiting for this
+// varible to be set and incremented when such a goroutine is
+// unblocked.
+type promise struct {
+	lock sync.Mutex
+
+	doneCh        chan struct{}
+	activeCounter counter.GoRoutineCounter // counter of active goroutines
+	waitingCount  int                      // number of goroutines idle due to this being unset
+
+	isSet         bool
+	value         interface{}
+}
+
+var _ WriteOnce = &promise{}
+
+// NewWriteOnce makes a new promise.LockingWriteOnce
+func NewWriteOnce(activeCounter counter.GoRoutineCounter) WriteOnce {
+	return &promise{
+		doneCh:       make(chan struct{}),
+		activeCounter: activeCounter,
+	}
+}
+
+// FIXME:
+var timeoutErr = fmt.Errorf("timeout")
+
+func (p *promise) WaitAndGet(stopCh <-chan struct{}) (interface{}, error) {
+	isSet := func() bool {
+		p.lock.Lock()
+		defer p.lock.Unlock()
+		if p.isSet {
+			return true
+		}
+
+		p.waitingCount++
+		p.activeCounter.Add(-1)
+		return false
+	}()
+	if isSet {
+		return p.value, nil
+	}
+
+	select {
+	case <-p.doneCh:
+		return p.value, nil
+	case <-stopCh:
+		p.lock.Lock()
+		defer p.lock.Unlock()
+		// Check if in the meantime value wasn't set (and thus the goroutine wasn't activated).
+		select {
+		case <-p.doneCh:
+			return p.value, nil
+		default:
+		}
+		// Now we know we really timed out. So we reactivate the goroutine.
+		p.waitingCount--
```

Comment:
For this range of code, there a goroutine running it and there is no blockage in sight, but the goroutine is not counted in the GoRoutineCounter.  This goroutine will be incorrectly ignored by a concurrent goroutine in FakeEventClock::Run.  For counting without timing splinters, the pattern that needs to be followed is to increase the count _before_ unblocking some existing goroutine or starting a new one.

Reply:
And I do that - this is happening only in the case where I actually timeout, in which case there is no other way to count that [and this is compatible with how counting currently works]

Reply:
I suppose that by "and I do that" you mean https://github.com/kubernetes/kubernetes/blob/8bf5872d8da305b3987bd8344914769acb249d35/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go#L699-L700 .  That makes https://github.com/kubernetes/kubernetes/blob/8bf5872d8da305b3987bd8344914769acb249d35/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/promise/promise.go#L86 double-counting of the same goroutine.  One of those two has to go.

If we delete the former then the timing splinter I described is present.  If we delete the latter then the code is only correct if every closure of `Context.Done()` is preceded by a counter increment _and_ we know that in that case the promise does not resume until that closure.  This is true in the one test case that exercises closure currently, but it seems an unnecessarily fragile pattern to establish.  I would rather use the more robust pattern with a goroutine in the test code.  I appreciate the desire to remove unnecessary costs from the APF filter in the non-test code.  What I am proposing in #104002 is to use a goroutine only in the test code.